ffi/dotnet-idsdk/IoticsIdLib/src/Tools.cs

```diff
@@ -0,0 +1,116 @@
+namespace IOTICS;
+
+class Tools {
+    
+    
+    public static string InvokeGoFunction(Func<IdLib.Return> goFunction)
+    {
+        string? error;
+        IdLib.Return result;
+
+        try
+        {
+            // Call the Go function
+            result = goFunction.Invoke();
+        }
+        catch (Exception ex)
+        {
+            // Handle any unexpected exceptions
+            throw new InvalidOperationException($"Exception when invoking method: {ex.Message}");
+        }
+
+        // Check for errors in r1 and throw an exception if necessary
+        error = IdLib.PtrToStringAndFree(result.r1);
+        if (error != null)
+        {
+            throw new IoticsLibException(error);
+        }
+
+        // Convert r0 to a string and free memory
+        string? value = IdLib.PtrToStringAndFree(result.r0);
+
+        if (value == null)
+        {
+            throw new IoticsLibException("unexpected null result");
+        }
+        return value;
+    }
+    
+    public static string? InvokeGoFunction(Func<IntPtr> goFunction)
+    {
+        IntPtr result;
+
+        try
+        {
+            // Call the Go function
+            result = goFunction.Invoke();
+        }
+        catch (Exception ex)
+        {
+            // Handle any unexpected exceptions
+            throw new InvalidOperationException($"Exception when invoking method: {ex.Message}");
+        }
+
+        // Handle the case where the Go function returns nil
+        if (result == IntPtr.Zero)
+        {
+            return null;
+        }
+
+        // Convert r0 to a string and free memory
+        string? value = IdLib.PtrToStringAndFree(result);
+
+        if (value == null)
+        {
+            throw new IoticsLibException("unexpected null result");
+        }
+        return value;
+    }
+
+    // Check if the given string is a URL
+    public static bool IsUrl(string input)
+    {
+        return Uri.TryCreate(input, UriKind.Absolute, out _);
```

Comment:
I don't like this - every URL validator is different - I don't think it guarantees anything really. What is wrong with just letting the calls fail when an invalid URL is supplied?

Reply:
It's a security practice
https://owasp.deteact.com/cheat/cheatsheets/Input_Validation_Cheat_Sheet.html
> . Input validation should happen as early as possible in the data flow, preferably as soon as the data is received from the external party.